receiver/mongodbreceiver/scraper.go

```diff
@@ -0,0 +1,197 @@
+	logger    *zap.Logger
+	config    *Config
+	client    client
+	extractor *extractor
+}
+
+func newMongodbScraper(logger *zap.Logger, config *Config) *mongodbScraper {
+	clientLogger := logger.Named("mongo-scraper")
+	c := NewClient(config, clientLogger)
+
+	return &mongodbScraper{
+		logger: logger,
+		config: config,
+		client: c,
+	}
+}
+
+func (s *mongodbScraper) start(ctx context.Context, _ component.Host) error {
+	if err := s.client.Connect(ctx); err != nil {
+		return fmt.Errorf("unable to connect to mongo instance: %w", err)
+	}
+
+	vr, err := s.client.GetVersion(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to get a version from the mongo instance: %w", err)
+	}
+
+	s.extractor, err = newExtractor(*vr, s.logger)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *mongodbScraper) shutdown(ctx context.Context) error {
+	if s.client != nil {
+		return s.client.Disconnect(ctx)
+	}
+	return nil
+}
+
+func (s *mongodbScraper) scrape(ctx context.Context) (pdata.Metrics, error) {
+	s.logger.Debug("starting otelcol/mongodb scrape")
+	if s.client == nil {
+		return pdata.NewMetrics(), errors.New("no client was initialized before calling scrape")
+	}
+	return s.collectMetrics(ctx, s.client)
+}
+
+func (s *mongodbScraper) collectMetrics(ctx context.Context, client client) (pdata.Metrics, error) {
+	rms := pdata.NewMetrics()
+	ilm := rms.ResourceMetrics().AppendEmpty().InstrumentationLibraryMetrics().AppendEmpty()
+	ilm.InstrumentationLibrary().SetName("otelcol/mongodb")
+	mm := newMetricManager(s.logger, ilm)
+	dbNames, err := client.ListDatabaseNames(ctx, bson.D{})
+	if err != nil {
+		s.logger.Error("Failed to fetch database names", zap.Error(err))
+		return pdata.NewMetrics(), err
+	}
+
+	wg := &sync.WaitGroup{}
+	var errors scrapererror.ScrapeErrors
+
+	wg.Add(1)
+	go s.collectAdminDatabase(ctx, wg, mm, errors)
+
+	for _, dbName := range dbNames {
+		wg.Add(1)
+		go s.collectDatabase(ctx, wg, mm, dbName, errors)
+	}
+
+	wg.Wait()
+
+	return rms, errors.Combine()
+}
+
+func (s *mongodbScraper) collectDatabase(ctx context.Context, wg *sync.WaitGroup, mm *metricManager, databaseName string, errors scrapererror.ScrapeErrors) {
+	defer wg.Done()
+	dbStats, err := s.client.DBStats(ctx, databaseName)
+	if err != nil {
+		errors.AddPartial(1, err)
+	} else {
+		s.extractor.Extract(dbStats, mm, databaseName, normalDBStats)
+	}
+
+	serverStatus, err := s.client.ServerStatus(ctx, databaseName)
+	if err != nil {
+		errors.AddPartial(1, err)
+	} else {
+		s.extractor.Extract(serverStatus, mm, databaseName, normalServerStats)
+	}
+}
+
+func (s *mongodbScraper) collectAdminDatabase(ctx context.Context, wg *sync.WaitGroup, mm *metricManager, errors scrapererror.ScrapeErrors) {
+	defer wg.Done()
+	serverStatus, err := s.client.ServerStatus(ctx, "admin")
+	if err != nil {
+		errors.AddPartial(1, err)
+	} else {
+		s.extractor.Extract(serverStatus, mm, "admin", adminServerStats)
+	}
+}
+
+type metricManager struct {
```

Comment:
A similar struct has been added to the mdatagen package. Please take a look at how this is being done on the `elasticreceiver`, for example. We should not need to manage all of this in the scraper.